File: lemonade-idle/idle-guard.py
#!/usr/bin/env python3
"""
Exits 0 if Lemonade Server GPU has been below THRESHOLD% for the entire
sliding window. Exits 1 (busy) otherwise.

Usage: idle-guard.py [--threshold N] [--window-s N] [--min-samples N]
"""
import json
import os
import sys
import time
import argparse

STATS_FILE = os.environ.get("LEMONADE_STATS_FILE",
             os.path.expanduser("~/.local/share/lemonade-idle/gpu.jsonl"))

def main():
    p = argparse.ArgumentParser()
    p.add_argument("--threshold",   type=float, default=float(os.environ.get("GPU_THRESHOLD",   "25")))
    p.add_argument("--window-s",    type=int,   default=int(os.environ.get("WINDOW_S",          "900")))
    p.add_argument("--min-samples", type=int,   default=int(os.environ.get("MIN_SAMPLES",       "10")))
    args = p.parse_args()

    now    = int(time.time())
    cutoff = now - args.window_s

    if not os.path.exists(STATS_FILE):
        print("BUSY: no stats file yet — sampler not running?", file=sys.stderr)
        sys.exit(1)

    samples = []
    with open(STATS_FILE) as f:
        for line in f:
            try:
                e = json.loads(line)
                if e["ts"] >= cutoff:
                    samples.append(e)
            except Exception:
                pass

    if len(samples) < args.min_samples:
        print(f"BUSY: only {len(samples)} samples in window (need {args.min_samples}) — too early to decide", file=sys.stderr)
        sys.exit(1)

    peak = max(s["gpu"] for s in samples)
    if peak > args.threshold:
        print(f"BUSY: peak GPU was {peak:.1f}% in the last {args.window_s//60}min (threshold {args.threshold}%)", file=sys.stderr)
        sys.exit(1)

    avg = sum(s["gpu"] for s in samples) / len(samples)
    print(f"IDLE: {len(samples)} samples, peak {peak:.1f}%, avg {avg:.1f}% — below {args.threshold}%", file=sys.stderr)
    sys.exit(0)

if __name__ == "__main__":
    main()

File: agent-scheduler/crates/laio-common/src/idle_guard.rs
use anyhow::Result;
use chrono::Utc;
use serde::Deserialize;
use std::path::Path;

use crate::config::IdleGuardConfig;

#[derive(Debug, Deserialize)]
struct Sample {
    ts:  i64,
    gpu: f64,
}

#[derive(Debug, PartialEq)]
pub enum IdleStatus {
    Idle { samples: usize, peak_pct: f64, avg_pct: f64 },
    Busy { reason: String },
}

impl IdleStatus {
    pub fn is_idle(&self) -> bool {
        matches!(self, Self::Idle { .. })
    }
}

impl std::fmt::Display for IdleStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Idle { samples, peak_pct, avg_pct } =>
                write!(f, "IDLE ({samples} samples, peak {peak_pct:.1}%, avg {avg_pct:.1}%)"),
            Self::Busy { reason } =>
                write!(f, "BUSY: {reason}"),
        }
    }
}

pub fn check(cfg: &IdleGuardConfig) -> Result<IdleStatus> {
    if !cfg.enabled {
        return Ok(IdleStatus::Idle { samples: 0, peak_pct: 0.0, avg_pct: 0.0 });
    }

    let path = shellexpand::tilde(
        cfg.stats_file.to_str().unwrap_or("~/.local/share/lemonade-idle/gpu.jsonl")
    );
    let path = Path::new(path.as_ref());

    if !path.exists() {
        return Ok(IdleStatus::Busy {
            reason: format!("stats file not found: {} — is the sampler running?", path.display()),
        });
    }

    let content = std::fs::read_to_string(path)?;
    let cutoff  = Utc::now().timestamp() - cfg.window_s as i64;

    let samples: Vec<Sample> = content
        .lines()
        .filter(|l| !l.trim().is_empty())
        .filter_map(|l| serde_json::from_str::<Sample>(l).ok())
        .filter(|s| s.ts >= cutoff)
        .collect();

    if samples.len() < cfg.min_samples {
        return Ok(IdleStatus::Busy {
            reason: format!(
                "only {} samples in window (need {}) — sampler may have just started",
                samples.len(), cfg.min_samples
            ),
        });
    }

    let peak = samples.iter().map(|s| s.gpu).fold(f64::NEG_INFINITY, f64::max);
    let avg  = samples.iter().map(|s| s.gpu).sum::<f64>() / samples.len() as f64;

    if peak > cfg.threshold_pct {
        return Ok(IdleStatus::Busy {
            reason: format!(
                "peak GPU {peak:.1}% exceeded threshold {:.1}% in the last {}min",
                cfg.threshold_pct, cfg.window_s / 60
            ),
        });
    }

    Ok(IdleStatus::Idle { samples: samples.len(), peak_pct: peak, avg_pct: avg })
}
